sirius-pipeline/sirius_conntrack.p4

```diff
@@ -3,6 +3,140 @@
+          meta.conntrack_data.allow_in = true;
+  }
+
+  action conntrackIn_miss() {
+          if (hdr.tcp.flags == 0x2 /* SYN */) {
+            if (meta.direction == direction_t.OUTBOUND) {
+               add_entry("conntrackIn_allow"); // New PNA Extern
+               //adding failiure to be eventually handled
+               set_entry_expire_time(EXPIRE_TIME_PROFILE_LONG);
+            }
+          }
+  }
+
+  table conntrackIn {
+      key = {
+          directionNeutralAddr(meta.direction, hdr.ipv4.srcAddr, hdr.ipv4.dstAddr):
+              exact;
+          directionNeutralAddr(meta.direction, hdr.ipv4.dstAddr, hdr.ipv4.srcAddr):
+              exact;
+          hdr.ipv4.protocol : exact;
+          directionNeutralPort(meta.direction, hdr.tcp.srcPort, hdr.tcp.dstPort):
+              exact;
+          directionNeutralPort(meta.direction, hdr.tcp.dstPort, hdr.tcp.srcPort):
+              exact;
+      }
+      actions = {
+          conntrackIn_allow;
+          conntrackIn_miss;
+      }
+
+      add_on_miss = true; //New PNA property
+
+      idle_timeout_with_auto_delete = true; // New PNA property
+      const default_action = conntrackIn_miss; //New PNA property
+  }
+
+  apply{conntrackIn.apply();}
+}
+
+
+control ConntrackOut(inout headers_t hdr,
+                inout metadata_t meta)
+{
+  action conntrackOut_allow () {
+  /* Invalidate entry based on TCP flags */
+          if (hdr.tcp.flags & 0x101 /* FIN/RST */) {
+            set_entry_expire_time(EXPIRE_TIME_PROFILE_NOW); // New PNA extern
+            /* set entry to be purged */
+          }
+          restart_expire_timer(); // reset expiration timer for entry
+          meta.conntrack_data.allow_out = true;
+  }
+
+  action conntrackOut_miss() {
+          if (hdr.tcp.flags == 0x2 /* SYN */) {
+            if (meta.direction == direction_t.INBOUND) {
+               add_entry("ConntrackOut_allow"); // New PNA Extern
+               //adding failiure to be eventually handled
+               set_entry_expire_time(EXPIRE_TIME_PROFILE_LONG);
+            }
+          }
+  }
+
+  table conntrackOut {
+      key = {
+          directionNeutralAddr(meta.direction, hdr.ipv4.srcAddr, hdr.ipv4.dstAddr):
+              exact;
+          directionNeutralAddr(meta.direction, hdr.ipv4.dstAddr, hdr.ipv4.srcAddr):
+              exact;
+          hdr.ipv4.protocol : exact;
+          directionNeutralPort(meta.direction, hdr.tcp.srcPort, hdr.tcp.dstPort):
+              exact;
+          directionNeutralPort(meta.direction, hdr.tcp.dstPort, hdr.tcp.srcPort):
+              exact;
+          meta.eni : exact;
+      }
+      actions = {
+          conntrackOut_allow;
+          conntrackOut_miss;
+      }
+
+      add_on_miss = true; //New PNA property
+
+      idle_timeout_with_auto_delete = true; // New PNA property
+      const default_action = conntrackOut_miss; //New PNA property
+  }
+  apply {conntrackOut.apply();}
+}
+
+#endif // PNA_CONNTRACK
+
+
```

Comment:
I do not know whether this code matches the desired data plane behavior of the DASH features, but the code Mario has written here does appear to be compatible with the PNA specification in at least most ways.

The top level sirius_pipeline.p4 file still include the v1model.p4 include file used for the v1model P4 architecture.  There is no urgent need to remove that with this PR, but I suppose the idea will be to remove that if and when the rest of the code is updated for the PNA architecture, too?

Another basic question of mine is that I see two different P4 table definitions here: conntrackIn and conntrackOut.  Is the idea that these will be two independent add-on-miss tables, each of which is only applied for one direction of traffic, i.e. one is applied only for inbound traffic, and the other only for outbound?  Again, I do not know the desired DASH data plane features well enough to say whether there should be one conntrack table applied for packets in both directions, or two separate tables, but wanted to ask to see if there was some other English description of the desired functionality that I could compare this with.

Reply:
I am not sure either if this code matches the original behavior. It would be beneficial if the original authors could take a look. 
I have used two tables to reproduce the same behavior of the two state graphs in the original code. My understanding is that both tables (like the state graphs in the original code) are applied in each direction and they are used to avoid using the same table twice in the same pipeline (i.e., in two separate stages). Specifically, the sequence of operations should be the following: 
1) look up a flow to check whether it is in the flow table
2) if not found, check the ACLs
3) if flow is allowed, add flow to the flow table 
The existing code uses one table (say TableIn) in step 1 and the other table (say TableOut) in step 3.
On the returning packets, the tables are used in the opposite order: the flow is found in TableOut, the ACL is skipped and the flow is added in TableIn as well.
Thereafter, lookups in both tables result in a hit. 
Again, this is my understanding and it would be important that it were verified by the authors of the original code.

Agreed with v1model.p4
The code also still contains the list match kind.

Reply:
@marian-pritsak and DASH Community, may we review / comment WRT this PR please?

Reply:
Theoretically, we can use a single table. If we have to guess from the code, the reason for the symmetric key generation is to use a single table.